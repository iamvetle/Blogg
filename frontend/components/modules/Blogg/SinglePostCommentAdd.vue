<template>
    <form @submit.prevent="tryAddComment" data-test="comment-add-form">
        <div>
            <BaseButton text="Submit" class="bg-primary text-onPrimary p-1 rounded-md mb-2"
                data-test="submit_comment_button" />
            <div data-test="input_comment_text">
                <BaseTextareaInput v-model="textInput" />
            </div>
        </div>
    </form>
</template>

<script setup lang="ts">

/**
 * @component CommentForm
 * @description Allows users to add comments to a specific post.
 * 
 * @prop {number} postId - The ID of the post to which the comment is being added.
 *
 * @emits {Event} submit - Emits an event when the comment is successfully submitted.
 */
const props = defineProps({
    postId: Number,
});

const baseCommentURL = `http://localhost:8888/api/post/${props.postId}/add-comment/`;
const allPostCommentsURL = `http://localhost:8888/api/post/${props.postId}/comments/`;
const textInput = ref("");

/**
 * @method tryAddComment
 * @description Submits a new comment to the post. Clears the input after submission and refetches comments.
 * @async
 * @returns
 */
const tryAddComment = async () => {
    if (textInput.value.trim()) {
        const postBody = { post: props.postId, content: textInput.value };
        const response = await postCommentOnPost(baseCommentURL, postBody);
        textInput.value = "";

        const responseComment = await getSinglePostComments(allPostCommentsURL);
        if (!responseComment) {
            console.error("Error fetching post comments");
        }

        if (!response) {
            console.error("Error submitting comment");
        }
    }
};
</script>

<style scoped></style>
