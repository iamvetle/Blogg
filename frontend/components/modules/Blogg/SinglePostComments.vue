<template>
    <div v-if="postStore.allComments">
        <h2>Comments written: ({{ num_of_comments }})</h2>
        <div>
            <SinglePostCommentsList :comments="postStore.allComments" :post-author="post.author.username"/>
        </div>

        <hr>

        <SinglePostCommentAdd v-if="authStore.isAuthenticated" :post-id="post.id" />
    </div>
</template>

<script setup lang="ts">

/**
 * This component is part of the single post [id] page
 * It deals with the comments associated with the post and
 * other comments stuff
 */

const authStore = useAuthStore()

const postStore = usePostStore()

/**
 * Counts the number of comments - I do it this way for me to keep the reactivity
 */
const num_of_comments = computed(() => {
    return postStore.allComments?.length ?? 0

})

defineProps<{
    post:PostSingleType
}>();

</script>

<style scoped></style>