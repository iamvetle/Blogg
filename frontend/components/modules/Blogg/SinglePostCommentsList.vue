<template>
    <div v-if="comments?.length">
        <div v-for="(comment, index) of comments" :key="index">
            <SinglePostCommentSingle
            :comment="comment"
            :post-author="postAuthor"
            class="p-4 border-2 mt-2 rounded-md"
            />
        </div>
    </div>
    <div v-else>
        <p>No comments.</p>
    </div>
</template>

<script setup lang="ts">
/**
 * Loops through all comments passed down through props, and
 * then loops through them and renders them all.
 */

defineProps<{
    comments:CommentType[] | null;
    postAuthor:string;
}>()

</script>

<style scoped>

</style>