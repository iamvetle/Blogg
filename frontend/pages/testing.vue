<template>
  <div class="flex">

    <form @submit.prevent method="post" class="flex flex-col" enctype="multipart/form-data">

      <label for="name">Name:</label>
      <input type="text" name="name" id="name">

      <label for="email">Email</label>
      <input type="email" name="email" id="email">

      <label for="message">Message:</label>
      <textarea name="message" id="" cols="30" rows="10" placeholder="Write a message here"></textarea>

      <label for="file">File:</label>
      <input type="file" name="file" id="file">

      <button type="submit">Submit</button>
    </form>

  </div>
</template>

<script setup lang="ts">
onMounted(() => {

  const form = document.querySelector("form")

  form?.addEventListener("submit", (e: any) => {


    const formData = new FormData(form)

    for (let item of formData) {
      console.log(item[0], item[1])
    }
    
    // ta axios of prÃ¸v fetch her
  })





})

// watchEffect(() => {

// })

</script>

<style scoped></style>