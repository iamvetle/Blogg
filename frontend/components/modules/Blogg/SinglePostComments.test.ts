import SinglePostComments from './SinglePostComments.vue';
import { shallowMount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

import SinglePostCommentsList from '~/components/modules/Blogg/SinglePostCommentsList.vue';
import SinglePostCommentAdd from './SinglePostCommentAdd.vue';

let wrapper: any;
let pinia: any = createTestingPinia();

const mockPost = {
    id: 5,
    title: "testtitle",
    num_of_comments: 6
}

// let generalStore; 
// let postStore; 
// let loggedInUserStore; 
// let paginationStore; 

const factory = () => {
    return shallowMount(SinglePostComments, {
        global: {
            plugins: [pinia],
            components: {
                SinglePostCommentsList,
                SinglePostCommentAdd
            },
            mocks: {},
            stubs: {
                SinglePostCommentsList: true,
                SinglePostCommentAdd: true
            },
        },
        props: {
            post: mockPost
        },
        slots: {}
    })
};

describe('Testing the comments component that is inside of post id page', () => {

    beforeEach(() => {
        // generalStore = useGeneralStore(pinia); 
        // postStore = usePostStore(pinia); 
        // loggedInUserStore = useLoggedInUserStore(pinia); 
        // paginationStore = usePaginationStore(pinia); 

    });

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });

    test('Should exist', () => {
        wrapper = factory()

        expect(wrapper.exists()).toBe(true)
    })
    test('Should have a "comments" prop', () => {
        wrapper = factory()

        expect(wrapper.props("post")).toBeTruthy()
    })
    test('Should have the two comments components', () => {
        wrapper = factory()

        const commentList = wrapper.findComponent({ name: "SinglePostCommentsList" })
        expect(commentList.exists()).toBe(true)

        const commentAdd = wrapper.findComponent({ name: "SinglePostCommentAdd" })
        expect(commentAdd.exists()).toBe(true)
    })
    test('Should match snapshot', () => {
        wrapper = factory()

        expect(wrapper).toMatchSnapshot()
    })
});