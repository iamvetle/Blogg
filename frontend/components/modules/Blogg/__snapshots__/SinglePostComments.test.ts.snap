// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the comments component that is inside of post id page > Should match snapshot 1`] = `
"<div>
  <h2>Comments written: (0)</h2>
  <div>
    <single-post-comments-list-stub comments=\\"[object Object]\\" postauthor=\\"test_username\\"></single-post-comments-list-stub>
  </div>
  <hr>
  <single-post-comment-add-stub postid=\\"5\\"></single-post-comment-add-stub>
</div>"
`;
